torchvision/ops/dropblock.py

```diff
@@ -1 +1,64 @@
 import torch
+import torch.nn.functional as F
+from torch import nn, Tensor
+
+from ..utils import _log_api_usage_once
+
+
+class DropBlock2d(nn.Module):
+    """
+    Implements DropBlock2d from `"DropBlock: A regularization method for convolutional networks"
+    <https://arxiv.org/abs/1810.12890>`.
+
+    Args:
+        p (float): Probability of an element to be dropped.
+        block_size (int): Size of the block to drop.
+        inplace (bool): If set to ``True``, will do this operation in-place. Default: ``False``
+    """
+
+    def __init__(self, p: float, block_size: int, inplace: bool = False) -> None:
+        super(DropBlock2d, self).__init__()
+        _log_api_usage_once(self)
+
+        if p < 0.0 or p > 1.0:
+            raise ValueError(f"drop probability has to be between 0 and 1, but got {p}")
+        self.p = p
+        self.block_size = block_size
+        self.inplace = inplace
+
+    def forward(self, input: Tensor) -> Tensor:
+        """
+        Args:
+            input (Tensor): Input feature map on which some areas will be randomly
+                dropped.
+        Returns:
+            Tensor: The tensor after DropBlock layer.
+        """
+        if not self.training:
+            return input
+        
+        N, C, H, W = input.size()
+        # compute the gamma of Bernoulli distribution
+        gamma = (self.p * H * W) / ((self.block_size ** 2) * ((H - self.block_size + 1) * \
+            (W - self.block_size + 1)))
+        mask_shape = (N, C, H - self.block_size + 1, W - self.block_size + 1)
+        mask = torch.bernoulli(torch.full(mask_shape, gamma, device=input.device))
+
+        mask = F.pad(mask, [self.block_size // 2] * 4, value=0)
```

Comment:
Could you provide a reference for the choice of `mask_shape` and the subsequent padding with zeros that you do? I need to check the paper in more detail but my expectation was that we would sample from bernoulli with shape equal the input. I'll need to have a closer look on the paper to remember the details.

Reply:
@datumbox the implementation is consistent with the original paper, there is 3 steps:
1. generate the center mask, which means the block center, so the center mask shape : `mask_shape = (N, C, H - self.block_size + 1, W - self.block_size + 1)`; we also should convert the `drop_prob` to `gamma` of the bernoulli.
2. convert the center mask to block mask: `pad` and use a `maxpool` to get block mask.
3. use the block mask to mask the input.

Reply:
I've checked the paper in detail and you are right. Page 3, figure 2a shows that the center mask should be drawn from the green area, hence your approach is correct. Also this implementation aligns closely with the [original TF implementation](https://github.com/tensorflow/tpu/blob/a6005716b311f96190590414479f0261726270e7/models/official/resnet/resnet_model.py#L315-L316).